GoogleUtilities/Example/Tests/Swizzler/GULSceneDelegateSwizzlerTest.m

```diff
@@ -0,0 +1,153 @@
+    OCMStub([mockSharedScene delegate]).andReturn(realSceneDelegate);
+    size_t sizeBefore = class_getInstanceSize([GULTestSceneDelegate class]);
+
+    Class realSceneDelegateClassBefore = [realSceneDelegate class];
+
+    [GULSceneDelegateSwizzler proxySceneDelegateIfNeeded:mockSharedScene];
+
+    XCTAssertTrue([realSceneDelegate isKindOfClass:[GULTestSceneDelegate class]]);
+
+    NSString *newClassName = NSStringFromClass([realSceneDelegate class]);
+    XCTAssertTrue([newClassName hasPrefix:@"GUL_"]);
+    // It is no longer GULTestSceneDelegate class instance.
+    XCTAssertFalse([realSceneDelegate isMemberOfClass:[GULTestSceneDelegate class]]);
+
+    size_t sizeAfter = class_getInstanceSize([realSceneDelegate class]);
+
+    // Class size must stay the same.
+    XCTAssertEqual(sizeBefore, sizeAfter);
+
+    // After being proxied, it should be able to respond to the required method selector.
+    XCTAssertTrue([realSceneDelegate respondsToSelector:@selector(scene:openURLContexts:)]);
+
+    // Make sure that the class has changed.
+    XCTAssertNotEqualObjects([realSceneDelegate class], realSceneDelegateClassBefore);
+  }
+}
+
+- (void)testProxyProxiedSceneDelegate {
+  if (@available(iOS 13, tvOS 13, *)) {
+    GULTestSceneDelegate *realSceneDelegate = [[GULTestSceneDelegate alloc] init];
+    id mockSharedScene = OCMClassMock([UIScene class]);
+    OCMStub([mockSharedScene delegate]).andReturn(realSceneDelegate);
+
+    // Proxy the scene delegate for the 1st time.
+    [GULSceneDelegateSwizzler proxySceneDelegateIfNeeded:mockSharedScene];
+
+    Class realSceneDelegateClassBefore = [realSceneDelegate class];
+
+    // Proxy the scene delegate for the 2nd time.
+    [GULSceneDelegateSwizzler proxySceneDelegateIfNeeded:mockSharedScene];
+
+    // Make sure that the class isn't changed.
+    XCTAssertEqualObjects([realSceneDelegate class], realSceneDelegateClassBefore);
+  }
+}
+
+- (void)testSceneOpenURLContextsIsInvokedOnInterceptors {
+  if (@available(iOS 13, tvOS 13, *)) {
+    NSSet *urlContexts = [NSSet set];
+
+    GULTestSceneDelegate *realSceneDelegate = [[GULTestSceneDelegate alloc] init];
+    id mockSharedScene = OCMClassMock([UIScene class]);
+    OCMStub([mockSharedScene delegate]).andReturn(realSceneDelegate);
+
+    id interceptor = OCMProtocolMock(@protocol(TestSceneProtocol));
+    OCMExpect([interceptor scene:mockSharedScene openURLContexts:urlContexts]);
+
+    id interceptor2 = OCMProtocolMock(@protocol(TestSceneProtocol));
+    OCMExpect([interceptor2 scene:mockSharedScene openURLContexts:urlContexts]);
+
+    [GULSceneDelegateSwizzler proxySceneDelegateIfNeeded:mockSharedScene];
+
+    [GULSceneDelegateSwizzler registerSceneDelegateInterceptor:interceptor];
+    [GULSceneDelegateSwizzler registerSceneDelegateInterceptor:interceptor2];
+
+    [realSceneDelegate scene:mockSharedScene openURLContexts:urlContexts];
+    OCMVerifyAll(interceptor);
+    OCMVerifyAll(interceptor2);
+  }
+}
+
+#endif
+
```

Comment:
I think we have to copy the tests for honoring flags as well (https://github.com/firebase/firebase-ios-sdk/blob/6fe312c451f0afb66b1782b38d38ccc6f857575b/GoogleUtilities/Example/Tests/Swizzler/GULAppDelegateSwizzlerTest.m#L1110), unless we can move this logic to a separate place and test it there.

Reply:
Alex suggested that we reuse the same flags. What do you think?

Reply:
Do you mean using GULAppDelegateSwizzler to get value in the scene swizzler? If yes - SGTM.

Reply:
Reused AppDelegate flags.

Reply:
I think we still need tests for the logic protected by the flag even if we reuse the flag calculation itself.

Reply:
It makes sense. I gonna duplicate those tests of AppDelegateSwizzler to SceneDelegateSwizzler.